code/modules/reagents/chemistry/machinery/chem_recipe_debug.dm

```diff
@@ -105,25 +153,207 @@
+			continue
+		var/overheat = FALSE
+		var/danger = FALSE
+		var/purity_alert = 2 //same as flashing
+		if(reagent.purity < equilibrium.reaction.purity_min)
+			purity_alert = ENABLE_FLASHING//Because 0 is seen as null
+			danger = TRUE
+		if(!(flashing == ENABLE_FLASHING))//So that the pH meter flashes for ANY reactions out of optimal
+			if(equilibrium.reaction.optimal_ph_min > reagents.ph || equilibrium.reaction.optimal_ph_max < reagents.ph)
+				flashing = ENABLE_FLASHING
+		if(equilibrium.reaction.is_cold_recipe)
+			if(equilibrium.reaction.overheat_temp > reagents.chem_temp)
+				danger = TRUE
+				overheat = TRUE
+		else
+			if(equilibrium.reaction.overheat_temp < reagents.chem_temp)
+				danger = TRUE
+				overheat = TRUE
+		if(equilibrium.reaction.reaction_flags & REACTION_COMPETITIVE) //We have a compeitive reaction - concatenate the results for the different reactions
+			for(var/entry in active_reactions)
+				if(entry["name"] == reagent.name) //If we have multiple reaction methods for the same result - combine them
+					entry["reactedVol"] = equilibrium.reacted_vol
+					entry["targetVol"] = round(equilibrium.target_vol, 1)//Use the first result reagent to name the reaction detected
+					entry["quality"] = (entry["quality"] + equilibrium.reaction_quality) /2
+					continue
+		active_reactions.len++
+		active_reactions[length(active_reactions)] = list("name" = reagent.name, "danger" = danger, "purityAlert" = purity_alert, "quality" = equilibrium.reaction_quality, "overheat" = overheat, "inverse" = reagent.inverse_chem_val, "minPure" = equilibrium.reaction.purity_min, "reactedVol" = equilibrium.reacted_vol, "targetVol" = round(equilibrium.target_vol, 1))//Use the first result reagent to name the reaction detected
+	data["activeReactions"] = active_reactions
+	data["isFlashing"] = flashing
+
+	return data
+
+/obj/machinery/chem_recipe_debug/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	switch(action)
+		if("power")
+			return
+		if("temperature")
+			var/target = params["target"]
+			if(text2num(target) != null)
+				target = text2num(target)
+				. = TRUE
+			if(.)
+				f_temp = clamp(target, 0, 1000)
+		if("pH")
+			var/target = params["target"]
+			if(text2num(target) != null)
+				target = text2num(target)
+				. = TRUE
+			if(.)
+				f_ph = target
+		if("forceTemp")
+			force_temp = ! force_temp
+			. = TRUE
+		if("forcepH")
+			force_ph = ! force_ph
+			. = TRUE
+		if("react")
+			react = TRUE
+			return TRUE
+		if("all")
+			process_all = !process_all
+			return TRUE
+		if("beakerSpawn")
+			beaker_spawn = !beaker_spawn
+			return TRUE
+		if("setTargetList")
+			var/text = stripped_input(usr,"List","Enter a list of Recipe product names separated by commas", "Recipe", MAX_MESSAGE_LEN)
+			reaction_names = list()
+			if(!text)
+				say("Could not find reaction")
+			var/list/names = splittext("[text]", ",")
+			for(var/name in names)
+				var/datum/reagent/R = find_reagent_object_from_type(get_chem_id(name))
```

Comment:
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names